File: pillar-04-continuity/vantis-mobile/apps/ios/VantisMobile/Sources/VantisMobileFFI.swift
import Foundation

/// Swift wrapper for VantisMobile FFI bindings
/// Provides a safe, idiomatic Swift interface to the Rust library
public class VantisMobileFFI {
    
    // MARK: - Singleton
    
    public static let shared = VantisMobileFFI()
    
    private init() {
        vantis_initialize()
    }
    
    deinit {
        vantis_cleanup()
    }
    
    // MARK: - Version
    
    /// Get the library version
    public var version: String {
        var buffer = [CChar](repeating: 0, count: 64)
        let result = vantis_version(&buffer, UInt32(buffer.count))
        guard result.code == 0 else { return "unknown" }
        return String(cString: buffer)
    }
    
    // MARK: - KeyPair
    
    /// Wrapper for encryption key pair
    public class KeyPair {
        private var handle: OpaquePointer?
        
        public init?() {
            guard let ptr = vantis_keypair_generate() else {
                return nil
            }
            self.handle = OpaquePointer(ptr)
        }
        
        deinit {
            if let handle = handle {
                vantis_keypair_free(UnsafeMutableRawPointer(handle))
            }
        }
        
        /// Get public key as base64 encoded string
        public var publicKeyBase64: String {
            var buffer = [CChar](repeating: 0, count: 128)
            let result = vantis_keypair_public_key_base64(
                UnsafeRawPointer(handle),
                &buffer,
                UInt32(buffer.count)
            )
            guard result.code == 0 else { return "" }
            return String(cString: buffer)
        }
    }
    
    // MARK: - Encryptor
    
    /// Wrapper for encryption/decryption
    public class Encryptor {
        private var handle: OpaquePointer?
        
        /// Create an encryptor from a base64-encoded shared secret
        /// - Parameter sharedSecretBase64: Base64 encoded encryption key (32 bytes when decoded)
        public init?(sharedSecretBase64: String) {
            guard let ptr = sharedSecretBase64.withCString({ secretPtr in
                vantis_encryptor_create(secretPtr)
            }) else {
                return nil
            }
            self.handle = OpaquePointer(ptr)
        }
        
        deinit {
            if let handle = handle {
                vantis_encryptor_free(UnsafeMutableRawPointer(handle))
            }
        }
        
        /// Encrypt data and return JSON-encoded encrypted message
        /// - Parameter plaintext: Data to encrypt
        /// - Returns: JSON string containing encrypted message, or nil on error
        public func encrypt(_ plaintext: Data) -> String? {
            var jsonLen = UInt32(plaintext.count + 256) // Extra space for JSON overhead
            var jsonBuffer = [CChar](repeating: 0, count: Int(jsonLen))
            
            let result = plaintext.withUnsafeBytes { plaintextPtr in
                vantis_encrypt(
                    UnsafeMutableRawPointer(handle),
                    plaintextPtr.baseAddress?.assumingMemoryBound(to: CUnsignedChar.self),
                    UInt32(plaintext.count),
                    &jsonBuffer,
                    &jsonLen
                )
            }
            
            guard result == 0 else { return nil }
            return String(cString: jsonBuffer)
        }
        
        /// Decrypt a JSON-encoded encrypted message
        /// - Parameter jsonEncrypted: JSON string from encrypt()
        /// - Returns: Decrypted data, or nil on error
        public func decrypt(_ jsonEncrypted: String) -> Data? {
            var plaintextLen = UInt32(jsonEncrypted.count) // Usually enough
            var plaintext = [UInt8](repeating: 0, count: Int(plaintextLen))
            
            let result = jsonEncrypted.withCString { jsonPtr in
                vantis_decrypt(
                    UnsafeMutableRawPointer(handle),
                    jsonPtr,
                    &plaintext,
                    &plaintextLen
                )
            }
            
            guard result == 0 else { return nil }
            return Data(plaintext[0..<Int(plaintextLen)])
        }
    }
    
    // MARK: - Device Info
    
    /// Device types
    public enum DeviceType: Int {
        case ios = 0
        case android = 1
        case desktop = 2
        case laptop = 3
        case tablet = 4
    }
    
    /// Wrapper for device info
    public class DeviceInfo {
        private var handle: OpaquePointer?
        
        public init?(name: String, deviceType: DeviceType, osVersion: String, appVersion: String) {
            guard let ptr = name.withCString({ namePtr in
                osVersion.withCString { osPtr in
                    appVersion.withCString { appPtr in
                        vantis_device_info_create(
                            namePtr,
                            Int32(deviceType.rawValue),
                            osPtr,
                            appPtr
                        )
                    }
                }
            }) else {
                return nil
            }
            self.handle = OpaquePointer(ptr)
        }
        
        deinit {
            if let handle = handle {
                vantis_device_info_free(UnsafeMutableRawPointer(handle))
            }
        }
        
        /// Get device info as JSON
        public var json: String? {
            var buffer = [CChar](repeating: 0, count: 4096)
            let result = vantis_device_info_to_json(
                UnsafeRawPointer(handle),
                &buffer,
                UInt32(buffer.count)
            )
            guard result.code == 0 else { return nil }
            return String(cString: buffer)
        }
    }
    
    // MARK: - Protocol Messages
    
    /// Create a ping message
    public func createPingMessage() -> String? {
        var buffer = [CChar](repeating: 0, count: 4096)
        let result = vantis_message_ping(&buffer, UInt32(buffer.count))
        guard result.code == 0 else { return nil }
        return String(cString: buffer)
    }
    
    /// Create a sync request message
    public func createSyncRequestMessage(lastSyncTimestamp: UInt64) -> String? {
        var buffer = [CChar](repeating: 0, count: 4096)
        let result = vantis_message_sync_request(
            CUnsignedLong(lastSyncTimestamp),
            &buffer,
            UInt32(buffer.count)
        )
        guard result.code == 0 else { return nil }
        return String(cString: buffer)
    }
    
    /// Create a notification message
    public func createNotificationMessage(
        title: String,
        body: String,
        type: String,
        priority: Int
    ) -> String? {
        var buffer = [CChar](repeating: 0, count: 4096)
        let result = title.withCString { titlePtr in
            body.withCString { bodyPtr in
                type.withCString { typePtr in
                    vantis_message_notification(
                        titlePtr,
                        bodyPtr,
                        typePtr,
                        Int32(priority),
                        &buffer,
                        UInt32(buffer.count)
                    )
                }
            }
        }
        guard result.code == 0 else { return nil }
        return String(cString: buffer)
    }
}

// MARK: - Error Types

public enum VantisError: Error {
    case nullPointer
    case invalidUtf8
    case invalidData
    case encryptionError
    case decryptionError
    case connectionError
    case invalidState
    case outOfMemory
    case unknown(String)
    
    init(code: Int, message: String? = nil) {
        switch code {
        case 1: self = .nullPointer
        case 2: self = .invalidUtf8
        case 3: self = .invalidData
        case 4: self = .encryptionError
        case 5: self = .decryptionError
        case 6: self = .connectionError
        case 7: self = .invalidState
        case 8: self = .outOfMemory
        default: self = .unknown(message ?? "Unknown error")
        }
    }
}

File: pillar-04-continuity/vantis-mobile/apps/android/VantisMobile/ffi/src/main/java/com/vantiscorp/vantismobile/ffi/VantisMobileFFI.kt
package com.vantiscorp.vantismobile.ffi

/**
 * Kotlin JNI wrapper for VantisMobile FFI bindings
 * Provides a safe, idiomatic Kotlin interface to the Rust library
 */
object VantisMobileFFI {
    
    init {
        System.loadLibrary("vantis_mobile")
        initialize()
    }
    
    // MARK: - Native Methods
    
    /**
     * Initialize the library.
     * Called automatically when the class is loaded.
     */
    private external fun initialize()
    
    /**
     * Cleanup the library.
     * Should be called when the application is terminating.
     */
    external fun cleanup()
    
    /**
     * Get the library version.
     * @return Version string
     */
    external fun getVersion(): String
    
    // MARK: - KeyPair
    
    /**
     * Generate a new encryption key pair.
     * @return Native handle to the key pair (must be freed with keypairFree)
     */
    external fun keypairGenerate(): Long
    
    /**
     * Free a key pair.
     * @param handle Native handle from keypairGenerate()
     */
    external fun keypairFree(handle: Long)
    
    /**
     * Get public key as base64 encoded string.
     * @param handle Native handle from keypairGenerate()
     * @return Base64 encoded public key
     */
    external fun keypairPublicKeyBase64(handle: Long): String
    
    // MARK: - Encryptor
    
    /**
     * Create an encryptor from a base64-encoded shared secret.
     * @param sharedSecretBase64 Base64 encoded encryption key (32 bytes when decoded)
     * @return Native handle to the encryptor (must be freed with encryptorFree)
     */
    external fun encryptorCreate(sharedSecretBase64: String): Long
    
    /**
     * Free an encryptor.
     * @param handle Native handle from encryptorCreate()
     */
    external fun encryptorFree(handle: Long)
    
    /**
     * Encrypt data and return JSON-encoded encrypted message.
     * @param handle Native handle from encryptorCreate()
     * @param plaintext Data to encrypt
     * @return JSON string containing encrypted message
     */
    external fun encrypt(handle: Long, plaintext: ByteArray): String?
    
    /**
     * Decrypt a JSON-encoded encrypted message.
     * @param handle Native handle from encryptorCreate()
     * @param jsonEncrypted JSON string from encrypt()
     * @return Decrypted data
     */
    external fun decrypt(handle: Long, jsonEncrypted: String): ByteArray?
    
    // MARK: - Device Info
    
    /**
     * Create device info.
     * @param name Device name
     * @param deviceType Device type (0=iOS, 1=Android, 2=Desktop, 3=Laptop, 4=Tablet)
     * @param osVersion OS version string
     * @param appVersion App version string
     * @return Native handle to device info (must be freed with deviceInfoFree)
     */
    external fun deviceInfoCreate(
        name: String,
        deviceType: Int,
        osVersion: String,
        appVersion: String
    ): Long
    
    /**
     * Free device info.
     * @param handle Native handle from deviceInfoCreate()
     */
    external fun deviceInfoFree(handle: Long)
    
    /**
     * Get device info as JSON.
     * @param handle Native handle from deviceInfoCreate()
     * @return JSON string
     */
    external fun deviceInfoToJson(handle: Long): String?
    
    // MARK: - Protocol Messages
    
    /**
     * Create a ping message.
     * @return JSON message string
     */
    external fun messagePing(): String?
    
    /**
     * Create a sync request message.
     * @param lastSyncTimestamp Last sync timestamp (milliseconds since epoch)
     * @return JSON message string
     */
    external fun messageSyncRequest(lastSyncTimestamp: Long): String?
    
    /**
     * Create a notification message.
     * @param title Notification title
     * @param body Notification body
     * @param type Notification type
     * @param priority Priority (0=low, 1=normal, 2=high, 3=urgent)
     * @return JSON message string
     */
    external fun messageNotification(
        title: String,
        body: String,
        type: String,
        priority: Int
    ): String?
}

// MARK: - Wrapper Classes

/**
 * Wrapper for encryption key pair
 */
class VantisKeyPair {
    private var handle: Long = 0
    
    init {
        handle = VantisMobileFFI.keypairGenerate()
        if (handle == 0L) {
            throw VantisError.InvalidData("Failed to generate key pair")
        }
    }
    
    /**
     * Get public key as base64 encoded string
     */
    val publicKeyBase64: String
        get() = VantisMobileFFI.keypairPublicKeyBase64(handle)
    
    /**
     * Free the key pair resources
     */
    fun free() {
        if (handle != 0L) {
            VantisMobileFFI.keypairFree(handle)
            handle = 0
        }
    }
    
    protected fun finalize() {
        free()
    }
}

/**
 * Wrapper for encryption/decryption
 */
class VantisEncryptor(sharedSecretBase64: String) {
    private var handle: Long = 0
    
    init {
        handle = VantisMobileFFI.encryptorCreate(sharedSecretBase64)
        if (handle == 0L) {
            throw VantisError.InvalidData("Failed to create encryptor")
        }
    }
    
    /**
     * Encrypt data and return JSON-encoded encrypted message
     * @param plaintext Data to encrypt
     * @return JSON string containing encrypted message, or null on error
     */
    fun encrypt(plaintext: ByteArray): String? {
        return VantisMobileFFI.encrypt(handle, plaintext)
    }
    
    /**
     * Decrypt a JSON-encoded encrypted message
     * @param jsonEncrypted JSON string from encrypt()
     * @return Decrypted data, or null on error
     */
    fun decrypt(jsonEncrypted: String): ByteArray? {
        return VantisMobileFFI.decrypt(handle, jsonEncrypted)
    }
    
    /**
     * Free the encryptor resources
     */
    fun free() {
        if (handle != 0L) {
            VantisMobileFFI.encryptorFree(handle)
            handle = 0
        }
    }
    
    protected fun finalize() {
        free()
    }
}

/**
 * Device types
 */
enum class VantisDeviceType(val value: Int) {
    IOS(0),
    ANDROID(1),
    DESKTOP(2),
    LAPTOP(3),
    TABLET(4)
}

/**
 * Wrapper for device info
 */
class VantisDeviceInfo(
    name: String,
    deviceType: VantisDeviceType,
    osVersion: String,
    appVersion: String
) {
    private var handle: Long = 0
    
    init {
        handle = VantisMobileFFI.deviceInfoCreate(
            name,
            deviceType.value,
            osVersion,
            appVersion
        )
        if (handle == 0L) {
            throw VantisError.InvalidData("Failed to create device info")
        }
    }
    
    /**
     * Get device info as JSON
     */
    val json: String?
        get() = VantisMobileFFI.deviceInfoToJson(handle)
    
    /**
     * Free the device info resources
     */
    fun free() {
        if (handle != 0L) {
            VantisMobileFFI.deviceInfoFree(handle)
            handle = 0
        }
    }
    
    protected fun finalize() {
        free()
    }
}

// MARK: - Error Types

sealed class VantisError(message: String) : Exception(message) {
    class NullPointer : VantisError("Null pointer")
    class InvalidUtf8 : VantisError("Invalid UTF-8")
    class InvalidData(message: String) : VantisError(message)
    class EncryptionError(message: String) : VantisError(message)
    class DecryptionError(message: String) : VantisError(message)
    class ConnectionError(message: String) : VantisError(message)
    class InvalidState(message: String) : VantisError(message)
    class OutOfMemory : VantisError("Out of memory")
    class Unknown(message: String) : VantisError(message)
    
    companion object {
        fun fromCode(code: Int, message: String? = null): VantisError {
            return when (code) {
                1 -> NullPointer()
                2 -> InvalidUtf8()
                3 -> InvalidData(message ?: "Invalid data")
                4 -> EncryptionError(message ?: "Encryption error")
                5 -> DecryptionError(message ?: "Decryption error")
                6 -> ConnectionError(message ?: "Connection error")
                7 -> InvalidState(message ?: "Invalid state")
                8 -> OutOfMemory()
                else -> Unknown(message ?: "Unknown error")
            }
        }
    }
}